src/services/hsm.service.js

```diff
@@ -6,16 +6,210 @@
+  const levels = [
+    () => keyname, // Root level
+    name => `${name}_bip44`,
+    name => `${name}_coinHTR`,
+    name => `${name}_htrAcct`,
+    name => `${name}_htrChange0`,
+    /**
+     * This unconventional naming is due to the 32 chars keyname restriction,
+     * we reserve 17 chars for context of which "*_HAddr_*" uses 7, leaving 10
+     * for the address index, this allows us to use all 2 billion indexes
+     * without resorting to hex conversion since 2 billion in base 10 has
+     * length 10.
+     */
+    (name, opts) => `${name}_HAddr_${opts.index}`,
+  ];
+
+  return levels[pathIndex](keyname, options);
+}
+
+function getAcctKeyname(keyname) {
+  return getBip32Keyname(keyname, 3);
+}
+
+function getChangeKeyname(keyname) {
+  return getBip32Keyname(keyname, 4);
+}
+
+/**
+ * @typedef {Object} SignatureData
+ * @property {Buffer} signature
+ * @property {Buffer} pubkey
+ * @property {number} addressIndex
+ * @property {number} inputIndex
+ */
+
+class HsmSession {
+  /**
+   * @param {hsm.interfaces.Hsm} conn
+   * @param {string} rootKeyname
+   */
+  constructor(conn, rootKeyname) {
+    this.conn = conn;
+    this.rootKey = rootKeyname;
+    /**
+     * @type {Record<number, string>}
+     */
+    this.addressKeys = {};
+  }
+
+  /**
+   * Get the adddress index keyname from the address index
+   * @param {index} index - bip32 Address index
+   * @returns {Promise<string>}
+   */
+  async getAddressKeyName(index) {
+    if (index in this.addressKeys) {
+      return this.addressKeys[index];
+    }
+
+    const addressKeyName = getBip32Keyname(this.rootKey, 5, { index });
+    const changeKeyName = getChangeKeyname(this.rootKey);
+    try {
+      await this.conn.blockchain.createBip32ChildKeyDerivation(
+        hsm.enums.VERSION_OPTIONS.BIP32_HTR_MAIN_NET,
+        index,
+        false,
+        true,
+        changeKeyName,
+        addressKeyName,
+      );
+    } catch (err) /* istanbul ignore next */ {
+      // After an error, the HSM lib has a runtime issue that will fail the next call
+      // To avoid this we can wait 500ms before continuing.
+      // This may be fixed in the near future.
+      await delay(500);
+      // (CODE: 5022 | NAME: ERR_OBJ_ALREADY_EXISTS)
+      // This error code means the address key was already derived
+      if (err.errorCode === hsm.constants.ERR_OBJ_ALREADY_EXISTS) {
+        // address key already exists, just return it
+        this.addressKeys[index] = addressKeyName;
+        return addressKeyName;
+      }
+      console.error(err);
+      throw err;
+    }
+
+    this.addressKeys[index] = addressKeyName;
+    return addressKeyName;
+  }
+
+  /**
+   * Sign the given data with the given HSM key
+   * @param {HsmSession} hsmSession
+   * @param {Buffer} dataToSignHash
+   * @param {number} addressIndex
+   * @returns {Promise<Buffer>} DER encoded signature
+   */
+  async signData(dataToSignHash, addressIndex) {
```

Comment:
Maybe change the name to `getSignature`, since you are not really signing anything here. What do you think?

Reply:
Done